k8s: Split SharedResources into binary specific cells

pippolo84 opened the pull request:
Split the SharedResources cell into multiple cells, one for each binary (agent, operator and clustermesh-apiserver).
This improves the dependencies management, since each binary will depend only on the k8s resources that it actually uses.

Also, the functional options pattern is introduced to give more flexibility when customizing a resource events stream.

_Notes for the reviewers_: review commit by commit might be simpler. The last commit is the one responsible for the high number of files modified, but it is just a rename (`sharedResources` -> `resources`) after the removal of the single SharedResources cell.

Fixes https://github.com/cilium/cilium/issues/24070

joamaki commented on pkg/k8s/resource_ctors.go:
```diff
@@ -0,0 +1,120 @@
+	return resource.New[*cilium_api_v2.CiliumIdentity](lc, lw), nil
+}
+
+func CiliumNetworkPolicy(lc hive.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) (resource.Resource[*cilium_api_v2.CiliumNetworkPolicy], error) {
```
I wonder if it'd make sense to namespace these by package, e.g. `pkg/k8s/resources` and then we'd have `resources.CiliumNetworkPolicy`, or by the suffix (`NamespaceResource`). I think either is probably fine and I do like having `k8s` in the package name. If you go the suffix route, then let's add it to these `Cilium*` functions here.

pippolo84 replied:
Added the `Resource` suffix where missing.

pippolo84 committed:
k8s: Add Resource suffix to k8s resources constructors

Add the "Resource" suffix to the CNP, CCNP and CCG resources
constructors to uniform the methods naming.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
k8s: Use functional options for k8s resources ctors

Use functional options pattern to customize k8s resources constructors
with option modifiers.  This pattern allows to expose a base constructor
that is easily customizable with modifiers like labels-based objects
filtering.

This is helpful to customize the events stream for a given resource that
is used by multiple binaries, like the agent and the operator.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
k8s: Split SharedResources into binary specific cells

Split SharedResources cell into multiple cells, one for each binary
relying on k8s resources (agent, operator and clustermesh-apiserver).
Each binary depends only on the specific resources that it actually
uses, and it customizes each resource events stream based on its own
specific needs.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
k8s: Rename daemon SharedResources in Resources

The SharedResources cell has been split into multiple binary-specific
cells, so change the name of the daemon params and k8s watchers fields
accordingly.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

gandro reviewed (approved):
My codeowners look good to me.

It might be a naive question, but why do we even need the per-binary packages? Can't there be something like `cell.LazyProvide` that only executes the constructor if it's actually requested by a dependency?

YutaroHayakawa commented on clustermesh-apiserver/k8s/resources.go:
```diff
@@ -0,0 +1,25 @@
+		"Clustermesh-apiserver Kubernetes resources",
+
+		cell.Provide(
+			k8s.ServiceResource,
```
I think we need CiliumNode, Endpoints, and CiliumEndpoint resources, but they are currently using the legacy `clientset`. Approving with the understanding that they will be followed up at some point.

pippolo84 replied:
Yep, there is already work in progress to port other resources to the new `resource.Resource[T]` framework.

YutaroHayakawa reviewed (approved):
Looks good to me from ClusterMesh's perspective.

ldelossa reviewed (approved):
LGTM
